src/utility/Albany_ThyraCrsMatrixFactory.cpp

```diff
@@ -131,30 +139,82 @@ void ThyraCrsMatrixFactory::insertGlobalIndices (const GO row, const Teuchos::Ar
       }
       m_graph->e_graph->InsertGlobalIndices(e_row, e_size,e_indices.getRawPtr());
     }
+    */
 #else
     TEUCHOS_TEST_FOR_EXCEPTION (true, std::logic_error, "Error! Epetra is not enabled in albany.\n");
 #endif
   } else {
+    int lrow = t_range->getLocalElement(static_cast<Tpetra_GO>(row));
+    auto& row_indices = t_local_graph[lrow];
+    //row_indices.reserve(row_indices.size()+indices.size());
+    for (int i=0; i<indices.size(); ++i) {
+      ALBANY_EXPECT(indices[i]<=max_safe_gid, "Error! Input gids exceed Epetra_GO ranges.\n");
+      row_indices.emplace(static_cast<Tpetra_GO>(indices[i]));
+    }
+/*
     // Despite being both 64 bits, GO and Tpetra_GO *may* be different *types*.
     Teuchos::ArrayView<const Tpetra_GO> t_indices(reinterpret_cast<const Tpetra_GO*>(indices.getRawPtr()),indices.size());
     m_graph->t_graph->insertGlobalIndices(static_cast<Tpetra_GO>(row),t_indices);
+    */
   }
 }
 
 void ThyraCrsMatrixFactory::fillComplete () {
+
+
+
   auto bt = Albany::build_type();
   if (bt==BuildType::Epetra) {
 #ifdef ALBANY_EPETRA
+    Teuchos::ArrayRCP<int> nonzeros_per_row_array(e_range->NumMyElements());
+    for (int lrow=0; lrow<nonzeros_per_row_array.size(); ++lrow)
+      nonzeros_per_row_array[lrow] = e_local_graph[lrow].size();
+
+    m_graph->e_graph = Teuchos::rcp(new Epetra_CrsGraph(Copy,*e_range,nonzeros_per_row_array.getRawPtr(),true));
+
+    for (int lrow=0; lrow<nonzeros_per_row_array.size(); ++lrow) {
+      auto& row_indices = e_local_graph[lrow];
+      if(row_indices.size()>0) {
+        Teuchos::Array<Epetra_GO> e_indices(row_indices.size());
+        int i=0;
+        for (const auto &index : row_indices)
+          e_indices[i++] = index;
+        auto row = e_range->GID(lrow);
+        m_graph->e_graph->InsertGlobalIndices(row,e_indices.size(),e_indices.getRawPtr());
+      }
+    }
+
     auto e_domain = getEpetraBlockMap(m_domain_vs);
-    auto e_range  = getEpetraBlockMap(m_range_vs);
     m_graph->e_graph->FillComplete(*e_domain,*e_range);
     m_graph->e_graph->OptimizeStorage();
```

Comment:
Don't forget to clear the local graph storage.

Reply:
I cleared the local graph storage. The only things I did not clear are the range maps.

Reply:
Here https://github.com/SNLComputation/Albany/pull/529/files#diff-1cde7201d33ac3652684350f23422d55R189

Reply:
Thanks. For some reason I did not see it.

Reply:
The range maps are locals though, right? So they'll cleaned up upon return.

Reply:
They are not local, otherwise one would have to create them each time insertGlobalIndices is called. I'll clear them.

Reply:
Ah, right. Well, they are not a storage overhead, assuming the input vector spaces were storing an epetra block map in their extra data section, which should always be the case (except for the replicated vs corresponding to multi-vector domains, but that's not the case here).